### ISSUE TITLE
Using refine-arithmetic incorrectly succeeds 

### ISSUE THREAD

**Issue by xXMarkuski** [diffblue/cbmc#6322]
CBMC version: Tested with cbmc-5.37.0
Operating system: Fedora 34
Test case:
```c
#include <assert.h>

void main() {
  assert(0);
  int arr[2];

  arr[0] = 1;
  int i = 0;
  if (0 < arr[1])
    i = 1;
  int a = arr[1] * arr[i];
  assert(0);
}
```
Exact command line resulting in the issue: 
```bash
$cbmc testcase.c --refine-arithmetic | tail
Runtime Solver: 0.000147571s
Runtime decision procedure: 0.000839873s

** Results:
testcase.c function main
[main.assertion.1] line 4 assertion 0: SUCCESS
[main.assertion.2] line 12 assertion 0: SUCCESS

** 0 of 2 failed (1 iterations)
VERIFICATION SUCCESSFUL
```
What behaviour did you expect: The assertions fail
What happened instead: The assertions are successfully verified

Without ```--refine-arithmetic``` the verification fails correctly, cbmc-5.35.0 doesn't  have this issue. I did not test version 5.36.0.

**Comment by TGWDB** [comment]
Thanks for the report, I've tried this in `cbmc` version 5.37.0 (on Ubuntu) and have not been able to reproduce.

I'll see if I can set up a Fedora system, but commenting here in case anyone else can reproduce (or show that this does not hold). Note that I also tried with the current release version (5.38.0 with the same results).

```
tgw@DB0907:~/github/testing/Issue6322$ cat -n testcase.c
     1  #include <assert.h>
     2
     3  void main() {
     4    assert(0);
     5    int arr[2];
     6
     7    arr[0] = 1;
     8    int i = 0;
     9    if (0 < arr[1])
    10      i = 1;
    11    int a = arr[1] * arr[i];
    12    assert(0);
    13  }
tgw@DB0907:~/github/testing/Issue6322$ cbmc --version
5.37.0 (cbmc-5.37.0)
tgw@DB0907:~/github/testing/Issue6322$ cbmc testcase.c
CBMC version 5.37.0 (cbmc-5.37.0) 64-bit x86_64 linux
Parsing testcase.c
Converting
Type-checking testcase
Generating GOTO Program
Adding CPROVER library (x86_64)
Removal of function pointers and virtual functions
Generic Property Instrumentation
Running with 8 object bits, 56 offset bits (default)
Starting Bounded Model Checking
Runtime Symex: 0.0012798s
size of program expression: 50 steps
simple slicing removed 2 assignments
Generated 2 VCC(s), 2 remaining after simplification
Runtime Postprocess Equation: 1.5e-05s
Passing problem to propositional reduction
converting SSA
Runtime Convert SSA: 0.0031733s
Running propositional reduction
Post-processing
Runtime Post-process: 3.68e-05s
Solving with MiniSAT 2.2.1 with simplifier
2089 variables, 9707 clauses
SAT checker: instance is SATISFIABLE
Runtime Solver: 0.0054211s
Runtime decision procedure: 0.0086858s

** Results:
testcase.c function main
[main.assertion.1] line 4 assertion 0: FAILURE
[main.assertion.2] line 12 assertion 0: FAILURE

** 2 of 2 failed (2 iterations)
VERIFICATION FAILED
tgw@DB0907:~/github/testing/Issue6322$ cbmc testcase.c --refine-arithmetic
CBMC version 5.37.0 (cbmc-5.37.0) 64-bit x86_64 linux
Parsing testcase.c
Converting
Type-checking testcase
Generating GOTO Program
Adding CPROVER library (x86_64)
Removal of function pointers and virtual functions
Generic Property Instrumentation
Running with 8 object bits, 56 offset bits (default)
Starting Bounded Model Checking
Runtime Symex: 0.0019794s
size of program expression: 50 steps
simple slicing removed 2 assignments
Generated 2 VCC(s), 2 remaining after simplification
Runtime Postprocess Equation: 1.95e-05s
Passing problem to refinement loop with MiniSAT 2.2.1 with simplifier
converting SSA
Runtime Convert SSA: 0.0008442s
Running refinement loop with MiniSAT 2.2.1 with simplifier
BV-Refinement: post-processing
BV-Refinement: iteration 1
398 variables, 874 clauses
SAT checker: instance is UNSATISFIABLE
Found assumption for '0/*' in proof (state 0)
BV-Refinement: got UNSAT, and the proof fails, refining
BV-Refinement: iteration 2
398 variables, 461 clauses
SAT checker: instance is SATISFIABLE
BV-Refinement: got SAT, and it simulates => SAT
Total iterations: 2
Runtime Solver: 0.0013137s
Runtime decision procedure: 0.0022739s

** Results:
testcase.c function main
[main.assertion.1] line 4 assertion 0: FAILURE
[main.assertion.2] line 12 assertion 0: FAILURE

** 2 of 2 failed (2 iterations)
VERIFICATION FAILED
tgw@DB0907:~/github/testing/Issue6322$
```

**Comment by TGWDB** [comment]
I cannot reproduce on Fedora 34 (5.11.12-300.fc34.x86_64) with `cbmc` version 5.37.0.

```
[tgw@fedora Issue6322]$ uname -r
5.11.12-300.fc34.x86_64
[tgw@fedora Issue6322]$ cat -n  testcase.c 
     1	#include <assert.h>
     2	
     3	void main() {
     4	  assert(0);
     5	  int arr[2];
     6	
     7	  arr[0] = 1;
     8	  int i = 0;
     9	  if (0 < arr[1])
    10	    i = 1;
    11	  int a = arr[i] * arr[i];
    12	  assert(0);
    13	
    14	}
[tgw@fedora Issue6322]$ cbmc --version
5.37.0 (cbmc-5.37.0)
[tgw@fedora Issue6322]$ cbmc testcase.c | tail
Runtime Solver: 0.00416087s
Runtime decision procedure: 0.00725872s

** Results:
testcase.c function main
[main.assertion.1] line 4 assertion 0: FAILURE
[main.assertion.2] line 12 assertion 0: FAILURE

** 2 of 2 failed (2 iterations)
VERIFICATION FAILED
[tgw@fedora Issue6322]$ cbmc testcase.c --refine-arithmetic | tail
Runtime Solver: 0.00139615s
Runtime decision procedure: 0.00202373s

** Results:
testcase.c function main
[main.assertion.1] line 4 assertion 0: FAILURE
[main.assertion.2] line 12 assertion 0: FAILURE

** 2 of 2 failed (2 iterations)
VERIFICATION FAILED
[tgw@fedora Issue6322]$
```

**Comment by xXMarkuski** [comment]
Thanks for the reply, I forgot to mention I used CaDiCal over the IPASIR Interface (i.e. I built CMBC with `make -C src LIBS="$PWD/cadical/build/libcadical.a" IPASIR=$PWD/cadical/src`). I could verify the results you showed, when using the default custom interface.

**Comment by TGWDB** [comment]
Thanks for the update. I've experimented a bit and can confirm that I can (mostly) reproduce your outcome. However, for me this failure not only occurs in `cbmc` version `5.35.0` built with `CaDiCal` but also later versions (e.g. `5.38.0`). I did some further experiments and I'm pretty sure that issue is in `CaDiCal` (or at least the version used in `cbmc` using `IPASIR`) since using `IPASIR` with `Riss` does not have this problem (and `cbmc` version `5.35.0` does not have this problem with `Riss` (using `IPASIR`) or `MiniSAT`).
